File: library/disjoint_set.cpp
#include <bits/stdc++.h>
using namespace std;

struct DisjointSet {
    DisjointSet(int n) {
        for (int i = 0; i < n; ++i) {
            s.push_back(i);
            r.push_back(0);
        }
    }
    int find(int x) {
        if (s[x] != x) {
            s[x] = find(s[x]);
        }
        return s[x];
    }
    void join(int x, int y) {
        int sx = find(x);
        int sy = find(y);
        if (sx == sy) {
            return;
        }
        if (r[sx] < r[sy]) {
            swap(sx, sy);
        }
        s[sy] = sx;
        if (r[sx] == r[sy]) {
            r[sx] += 1;
        }
    }
    vector<int> s;
    vector<int> r;
};

File: library/disjoint_set.rs
struct DisjointSet {
    s: Vec<usize>,
    r: Vec<usize>,
}

impl DisjointSet {
    pub fn new(n: usize) -> Self {
        Self {
            s: (0..n).collect::<Vec<_>>(),
            r: vec![0; n]
        }
    }
    pub fn find(&mut self, x: usize) -> usize {
        if self.s[x] != x {
            self.s[x] = self.find(self.s[x]);
        }
        self.s[x]
    }
    pub fn union(&mut self, x: usize, y: usize) {
        let sx = self.find(x);
        let sy = self.find(y);
        if sx == sy {
            return;
        }
        let (sh, sl) = if self.r[sx] > self.r[sy] {
            (sx, sy)
        } else {
            (sy, sx)
        };
        self.s[sl] = sh;
        if self.r[sh] == self.r[sl] {
            self.r[sh] += 1;
        }
    }
}
